include/dlaf/common/round_robin.h

```diff
@@ -13,28 +13,46 @@
 
 #include <vector>
 
+#include <dlaf/common/assert.h>
+
 namespace dlaf {
 namespace common {
 
 template <class T>
-struct RoundRobin {
+class RoundRobin {
+public:
+  RoundRobin() = default;
+
   template <class... Args>
-  RoundRobin(std::size_t n, Args... args) : curr_index_(0) {
+  RoundRobin(std::size_t n, Args... args) {
     pool_.reserve(n);
     for (std::size_t i = 0; i < n; ++i)
       pool_.emplace_back(args...);
   }
 
+  RoundRobin(RoundRobin&&) = default;
+  RoundRobin(const RoundRobin&) = delete;
+  RoundRobin& operator=(RoundRobin&&) = default;
+  RoundRobin& operator=(const RoundRobin&) = delete;
+
   T& currentResource() {
+    DLAF_ASSERT(curr_index_ < pool_.size(), curr_index_, pool_.size());
```

Comment:
Updated this, https://github.com/eth-cscs/DLA-Future/pull/993/files/bb4b5967ea86368a1076aa7c8a1736408be802c6#diff-97018c62681fbe88215cbdd275c5177efbb4d7af6ed889fcd91cfbf11871a2d7, and the assertion for the `pool_` not being empty to `DLAF_ASSERT_HEAVY`. Or did you consciously not comment on `DLAF_ASSERT(!pool_.empty(), "")` because you'd prefer not to use `HEAVY` there?

Reply:
Updated this and the the other assertions in `round_robin.h` based on discussion. Summary: keep the first assertion in `currentResource` and `nextResource` as `DLAF_ASSERT` because they are runtime preconditions (the vector can be empty), but make the second, dependent, assertion in `nextResource` `DLAF_ASSERT_HEAVY` (because if the first assertion does not fail, the second should also only fail if the implementation is wrong, not by user error).